File: BaekJoon/DP/1로만들기_1463.swift
//
//  1로만들기_1463.swift
//  BackJoonPractice
//
//  Created by MinKyeongTae on 2020/02/27.
//  Copyright © 2020 Min Kyeong Tae. All rights reserved.
//

// MARK: - 1로만들기
// MARK: DP swift 문제풀이

/*
import Foundation

let N = Int(readLine()!)!

var DP = [Int](repeating: 0, count: 1000001)
DP[1] = 0
func go(_ N: Int) -> Int {
    if DP[N]>0 || N==1 { return DP[N] }
    
    var Ans = go(N-1) + 1
    if N % 3 == 0 {
        Ans = min(Ans, go(N/3) + 1)
    }
    if N % 2 == 0 {
        Ans = min(Ans, go(N/2) + 1)
    }
    DP[N] = Ans
    return DP[N]
}

print(go(N))
*/

File: BaekJoonPS/BaekjoonAlgorithmPS/DP/1로만들기_1463.cpp
//
//  1로만들기_1463.cpp
//  BackJoonPractice
//
//  Created by MinKyeongTae on 2019/12/09.
//  Copyright © 2019 Min Kyeong Tae. All rights reserved.
//

// MARK: - 1로 만들기

#if 0
#include <iostream>
#include <vector>
#include <cmath>
using namespace std;

vector<int> DP(1000001,0);
int go(int num) {
    if(num == 1) return 0;
    if(DP[num] > 0) return DP[num];
    DP[num] = go(num-1)+1;
    if(num%2==0) DP[num] = min(DP[num], go(num/2)+1);
    if(num%3==0) DP[num] = min(DP[num], go(num/3)+1);
    return DP[num];
}

int main() {
    ios_base::sync_with_stdio(0); cin.tie(0);
    int num; cin>>num;
    int Ans = go(num);
    printf("%d\n",Ans);
    return 0;
}
#endif
